### ISSUE TITLE
About GIoULoss.

### ISSUE THREAD

**Issue by StetchPlane** [kemaloksuz/aLRPLoss#9]
First of all, thanks for your excellent work!May I ask you to use "self.bbox_coder.encode()" when calculating bbox_targets, but instead of using "self.bbox_coder.decode()" when calculating pos_targets, you use your own "delta2bbox()", What is the reason for this?

https://github.com/kemaloksuz/aLRPLoss/blob/2cfaa8ec3150692cca8b085221f1bd8c8c78db8e/mmdet/models/dense_heads/alrp_loss_retina_head.py#L245

**Comment by kemaloksuz** [comment]
Hi,

Thanks for your interest in our work.

**Short answer:** Implementation simplicity and both of them should yield the same result. 

**Long answer if you are interested:**

We extend AnchorHead: 

https://github.com/kemaloksuz/aLRPLoss/blob/2cfaa8ec3150692cca8b085221f1bd8c8c78db8e/mmdet/models/dense_heads/alrp_loss_retina_head.py#L16

by overriding its loss() method:

https://github.com/kemaloksuz/aLRPLoss/blob/2cfaa8ec3150692cca8b085221f1bd8c8c78db8e/mmdet/models/dense_heads/alrp_loss_retina_head.py#L136

In aLRPRetinaHead, we do not use loss_single() of AnchorHead, which computes loss for each FPN layer. Instead, we collect all  inputs required to compute the loss()

https://github.com/kemaloksuz/aLRPLoss/blob/2cfaa8ec3150692cca8b085221f1bd8c8c78db8e/mmdet/models/dense_heads/alrp_loss_retina_head.py#L188

and send them to aLRP Loss computation at once, instead of FPN-layer wise loss computation. FPN-layer wise loss computation yields the same result for score based losses (e.g. cross-entropy and focal loss), however, for ranking-based losses using more examples provides a more accurate computation. This is why, we discard loss_single(). And unfortunately, this is the function in the base class, AnchorHead, where the boxes are decoded using "bbox_coder.decode()":

https://github.com/kemaloksuz/aLRPLoss/blob/2cfaa8ec3150692cca8b085221f1bd8c8c78db8e/mmdet/models/dense_heads/anchor_head.py#L413

The function bbox_coder.decode() requires two parameters anchors and offsets, and we did not put a lot of effort to organize the code to find appropriate (anchors, offsets) pairs. Namely, instead of trying to find which anchor correspond to which offset, we simply used an equivalent function (delta2bbox()), which has only a single parameter, the offsets.
